packages/svelte/src/runtime/internal/lifecycle.js

```diff
@@ -36,9 +42,7 @@ export function beforeUpdate(fn) {
  *
  * https://svelte.dev/docs#run-time-svelte-onmount
  * @template T
- * @param {() => T extends Promise<() => any>
- * 		? "Returning a function asynchronously from onMount won't call that function on destroy"
- * 		: T} fn
+ * @param {() => NotFunction<T> | Promise<NotFunction<T>> | (() => any)} fn
```

Comment:
I'm actually *surprised* this works as expected.

Consider that `NotFunction<T>` currently only negates `Function`. In theory, this should *not* exclude `Promise<T>`. Indeed, if we change the onMount type definition to:

```ts
function onMount<T>(
    fn: () => NotFunction<T> | (() => any)
): void
```

`fn` can now be anything, because the union of (not a function) and (a function) should logically be `any`. For some reason, a further union with `Promise<NotFunction<T>>` successfully narrows the type to exclude promises which return functions, and yet a standalone `Promise<NotFunction<T>>` type would block any non-async callback functions.

Is this behavior with TypeScript expected or a bug?
Even if it turns out to be a bug, we can instead narrow the conditional types:
```ts
type NotFunction<T> = T extends Function ? never : T;
type NotFunctionOrPromise<T> = T extends (Function | Promise<T>) ? never : T;
function onMount<T>(
    fn: () => NotFunctionOrPromise<T> | Promise<NotFunction<T>> | (() => any)
): void
```

Reply:
I think this is because the return type of the below code is `Promise<Fuinction>`.
So 

```ts
onMount(async () => {
	return () => {}
});
```

It means below code doesn't have a type error.

```ts
type NotFunction<T> = T extends Function ? never : T;
type A<T> = () => NotFunction<T>

function onMountA<T>(param: A<T>) {
    return param()
}

onMountA(async () => {
	return () => {}
});
```